claim spoofing fix

tgrecojs opened the pull request.

tgrecojs committed:
feat: codespace config

tgrecojs committed:
feat(depositFacet-claim): working implementation of using depositFacet to claim

tgrecojs committed:
chore: added depositFacet.test.js file

tgrecojs committed:
test(depositFacet): added test code for verifying that offers are being handled correctly when processed through updated offerHandler code

tgrecojs committed:
test: added code further usage of , and demonstrates that its well-suitedness as generic interface for creating AsyncGenerator subscribtions (#150)

This commit contains test for:
- Handling bobs offer
- Subscribing to bobs wallet
- Verifying Alice's inability to make a second claim attempt

tgrecojs committed:
test(claim-spoofing): this commit contains the changes  required for existing test code to properly evaluate assertions against updated makeClaimAirdropInvitation

- added Observable factory function for watching AsyncGenerators
- implemented logic for informing smart wallets of tribbles issuer (fixes issue where provisionSmartWallet fails when interfacing with a new issuer)
- added import statements for various type definitions (a practice that should continue whenever faced with VSCode intellisense errors stemming from unknown types)

amessbee commented:
@tgrecojs I am going to review this, hopefully, today but it looks good at first glance. Could you maybe add very basics details in readme regarding how to run this on local? Thanks.

amessbee commented on contract/Makefile:
```diff
@@ -13,8 +13,8 @@ list:
 balance-q:
-	kagd keys show user1 -a --keyring-backend="test"
-	kagd query bank balances $(ACCT_ADDR)
+	agd keys show tg -a --keyring-backend="test"
+	agd query bank balances $(ACCT_ADDR)
```
wondering why the change from `user1` to `tg`?

tgrecojs replied:
`tg` is short for thomas greco. this is the name i have associated with the main test account i use within the context of `agd`. 

In other words, there is no exciting / noteworthy reason behind the naming.

amessbee commented on contract/src/airdrop.contract.js:
```diff
@@ -367,21 +424,54 @@ export const start = async (zcf, privateArgs, baggage) => {
+              Object.values(payment).map(pmtP =>
+                E.when(pmtP, pmt => E(depositFacet).receive(pmt)),
+              ),
+            );
```
I was wondering if there is a case of multiple payments to an address given that `depositFacet` receives payment before the corresponding address is added to the `accountStore` in case of failure somewhere in between?

tgrecojs replied:
This is a great call out. Could you provide me with an example of one such error? 

I'm not necessarily worried about a failure happening, however your comments did introduce a new scenario to me in which a user makes subsequent offers to claim before this Promise has resolved. If the user is able to get passed the `accountStore` book keeping occurs, then they would seemingly be able to collect more than one payment. 

This is all the more reason for updating the code so that the book keeping logic happens before any `Payment` is sent out.

amessbee commented on contract/src/airdrop.contract.js:
```diff
@@ -367,21 +424,54 @@ export const start = async (zcf, privateArgs, baggage) => {
-              harden([
-                [tokenHolderSeat, claimSeat, { Tokens: paymentAmount }],
-                [claimSeat, tokenHolderSeat, { Fee: claimTokensFee }],
-              ]),
+              harden([[claimSeat, tokenHolderSeat, { Fee: claimTokensFee }]]),
             );
```
This probably none to very limit impact in the current context, but we are probably breaking offer-safety here by separating deposit of tokens from claiming Fee. WDYT?

tgrecojs replied:
We've opted to allow for the offer to occur in such a manner enforces offer-safety for the `tokenHolderSeat`. This is due to the nature of this specific offer as it uses `depositFacet`s one one side of the offer, and Zoe seat rearrange for the other. 

If this is an issue, then I think the fix would be to move the accountStore logic so that it occurs before the `depositFacet` is sent a payment, effectively stopping any instances in which a user manages to get passed the check on `accountStore` before the  `Promise.all` payment resolves. 

Does that make sense?

tgrecojs replied:
I've realized that a merge I conducted for adding [this commit](https://github.com/hindley-milner-systems/dapp-ertp-airdrop/pull/151/commits/51d3b7969e1015813144891007de3d94b4f0b18e) into the codebase got lost along the way. I'm going to address this today. I wanted to make note of it though as it relates to ensuring offer safety.

tgrecojs commented:
> @tgrecojs I am going to review this, hopefully, today but it looks good at first glance. Could you maybe add very basics details in readme regarding how to run this on local? Thanks.

Yes, no problem @amessbee. I will be sure to do so today.

Additionally, if you find that you would benefit from a deeper understanding of the code, I'd be more than happy to walk you through it during a short Zoom call. Feel free to reach out to me, and we can schedule a time that suits us both.

amessbee commented on contract/test/tribbles-airdrop/contract.test.js:
```diff
@@ -1,326 +1,641 @@
+});
+test.serial('makePauseContractInvitation', async t => {
+  const merkleRoot = merkleTreeAPI.generateMerkleRoot(
+    accounts.map(x => x.pubkey.key),
```
I see that all tests in this module are passing but if possible, can we refactor these two tests (`makePauseContractInvitation` and `E2E test`) - I kind of lost track halfway through :)

amessbee commented on ui/src/components/Orchestration/MakeOffer.tsx:
```diff
@@ -91,31 +93,32 @@ const proof =  [
-        }
-      };
+  const offerArgsValue = makeOfferArgs({
+    address: wallet.address,
```
Question: if we are not using address in the contract, why are we still sending it here or expecting in `offerArgs` - do we expect a future use case?

tgrecojs replied:
No, we're not expecting to use it ever. I didn't mean to include this file in the PR so sorry for the confusion.

amessbee commented:
> > @tgrecojs I am going to review this, hopefully, today but it looks good at first glance. Could you maybe add very basics details in readme regarding how to run this on local? Thanks.
> 
> Yes, no problem @amessbee. I will be sure to do so today.
> 
> Additionally, if you find that you would benefit from a deeper understanding of the code, I'd be more than happy to walk you through it during a short Zoom call. Feel free to reach out to me, and we can schedule a time that suits us both.

Awesome - that would be great. I think the contract code looks fine to me. I may still want to go through the tests with you if we can find time.

tgrecojs committed:
feat(offer-validation): added code for validating offerArgs data. **This commit contains code that was introduced to the codebase on a seperate branch**. This is meant to be as an acknowledgement of this undesired (yet somewhat necessary) process

tgrecojs committed:
test: updated makeOfferArgs function to produce an object with the correct shape

tgrecojs reviewed (commented):
> > > @tgrecojs I am going to review this, hopefully, today but it looks good at first glance. Could you maybe add very basics details in readme regarding how to run this on local? Thanks.
> > 
> > Yes, no problem @amessbee. I will be sure to do so today.
> > 
> > Additionally, if you find that you would benefit from a deeper understanding of the code, I'd be more than happy to walk you through it during a short Zoom call. Feel free to reach out to me, and we can schedule a time that suits us both.
> 
> Awesome - that would be great. I think the contract code looks fine to me. I may still want to go through the tests with you if we can find time. 

@amessbee, here is a link to schedule a meeting with me. If you're unable to find a time that suits you, please let me know. I'm more than happy to accommodate. :)

https://scheduler.zoom.us/t/s4gpdxv0/1hr

amessbee commented:
Thanks @tgrecojs - I have requested a meeting on Monday if that works. It may be too late though, so I am available at the same time on Fri/Sat/Sun so do let me know if we can meet earlier. Thanks.

tgrecojs commented:
Thank you for setting the meeting @amessbee. 

Considering the timeline, it would be really helpful if we could move the session to Saturday at the same time (9AM EST). This change would allow me to address any remaining action items before Monday, putting us in a strong position to finalize the review process at the start of the week.

If this new timing works for you, please let me know, and I'll adjust the invite accordingly. 

Thanks again!

amessbee commented:
@tgrecojs yes that works for me, let's meet on Saturday.

tgrecojs committed:
refactor(offerHandler): (#152)

* refactor(offerHandler):
	- moved offerHandler book keeping so that it takes place at the time when Payments are sent to claimant's depositFacets
	- removed instances where precomputed logic was being used once
	- removed address string from accountStore lookup message to eliminate unnecessary computation

tgrecojs committed:
chore(linting): removed unused variables

tgrecojs committed:
chore(linting): removed unused variables from airdrop.proposal.js

tgrecojs committed:
test:

- added test code demonstrating makePauseInvitation working as expected
- removed instances of unnecessary duplicate tests assertions
- renamed e2e.test.js to coreEval.test.js

tgrecojs commented:
hey @amessbee - can you stamp this as all good now that you have context into the decision to take this approach, and the conversations that were conducted with Dean, as well as others, confirming the validity of this approach? 

you may have noticed but everything you mentioned has been addressed. thanks for all the useful feedback btw 🙂

amessbee reviewed (approved):
Great work - everything looks good!
